### PR TITLE
JWT: Support multiple schemes in addition to Bearer

### LINKED ISSUE #397: JWT custom Authentication Scheme Name
Right now JWT authentication only handles "Bearer" authentication types:
https://github.com/ktorio/ktor/blob/4196d2818ccd5cdbafbd412f414c642b64ddc5fd/ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt#L111

It happens that instead of Bearer, in the realworld.io sample, it uses `Authentication: Token ...`
https://github.com/gothinkster/realworld/blob/master/api/Conduit.postman_collection.json#L203

I'm not sure if it is supported:
http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml

Still. Would be nice to allow to change the supported auth scheme or just load them all.

### PR DESCRIPTION
**Pull request by soywiz**
This should fix #397 required for supporting realworld with ktor https://github.com/gothinkster/node-express-realworld-example-app/blob/ba04b70c31af81ca7935096740a6e083563b3a4a/routes/auth.js#L5

### CODE REVIEW

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 1]
```diff
@@ -107,8 +116,8 @@ private suspend fun verifyAndValidate(call: ApplicationCall, jwtVerifier: JWTVer
     return validate(call, credentials)
 }
 
-private fun HttpAuthHeader?.getBlob() = when {
-    this is HttpAuthHeader.Single && authScheme == "Bearer" -> blob
+private fun HttpAuthHeader?.getBlob(supportedAuthSchemes: Set<String>) = when {
+    this is HttpAuthHeader.Single && authScheme in supportedAuthSchemes -> blob
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: should it be case sensitive?

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 2]
```diff
@@ -21,9 +21,17 @@ class JWTPrincipal(val payload: Payload) : Principal
...
     internal var authenticationFunction: suspend ApplicationCall.(JWTCredential) -> Principal? = { null }
 
+    internal var supportedAuthSchemes = setOf("Bearer")
     var realm: String = "Ktor Server"
     internal var verifier: ((HttpAuthHeader?) -> JWTVerifier?) = { null }
 
+    /**
+     * @param [supported] list of supported authentication schemes for JWT (by default only "Bearer")
+     */
+    fun authSchemes(vararg schemes: String = arrayOf("Bearer")) {
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: `require(schemes.isNotEmpty()) { " .... " }`

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt [hunk 3]
```diff
@@ -118,8 +127,8 @@ private fun ApplicationRequest.parseAuthorizationHeaderOrNull() = try {
     null
 }
 
-private fun HttpAuthHeader.Companion.bearerAuthChallenge(realm: String): HttpAuthHeader =
-        HttpAuthHeader.Parameterized("Bearer", mapOf(HttpAuthHeader.Parameters.Realm to realm))
+private fun HttpAuthHeader.Companion.bearerAuthChallenge(realm: String, supportedAuthSchemes: Set<String>): HttpAuthHeader =
+        HttpAuthHeader.Parameterized(supportedAuthSchemes.first(), mapOf(HttpAuthHeader.Parameters.Realm to realm))
```
Comment (cy6erGn0m) [inline ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: it looks like we need to produce multiple auth challenges, not just for the first one

File: ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt (hunk 1 above)
Comment (soywiz) [reply ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt]: Yes. It seems so: https://tools.ietf.org/html/rfc7235#section-2.1

**Comment by soywiz** [comment]
I have done some changes. Please re-review. 

> it looks like we need to produce multiple auth challenges, not just for the first one

I don't know about a reasonable way for providing different schemes there. So what I have done in the changes is to make one scheme the default one (the expected for this authentication instead of arbitrarily getting the first one of the set).

**Pull request merged**